vshard/storage/init.lua

```diff
@@ -2685,14 +2685,21 @@ end
 -- @retval nil, err Error.
 -- @retval values Success.
 local function storage_call(bucket_id, mode, name, args)
-    local ret1, ret2, ret3, _
-    local ok, err = bucket_ref(bucket_id, mode)
-    if not ok then
-        return ok, err
+    local ret1, ret2, ret3
+    local ok, ok_ref, err
+    ok_ref, err = bucket_ref(bucket_id, mode)
+    if not ok_ref then
+        return nil, err
     end
     ok, ret1, ret2, ret3 = local_call(name, args)
-    _, err = bucket_unref(bucket_id, mode)
-    assert(not err)
+    ok_ref, err = bucket_unref(bucket_id, mode)
+    if not ok_ref then
+        -- It should not normally happen. But the bucket could be deleted
+        -- manually or due to a bug. Then the read request could see
+        -- inconsistent data - the bucket could be half-deleted, for example.
+        -- Regardless of what user's function returns, treat it as an error.
+        return nil, err
+    end
```

Comment:
I propose to write log with the original error (ret1) if it happened else it will be lost. Or write it to log in any case.

Reply:
In the new version of the patch I added support for stacked diagnostics and return the user function's exception in `err.prev` member. The caller can now log it if he wants.